### PR TITLE
chore: bump google cloud storage

### LINKED ISSUE #3668: update `google.golang.org/cloud`
Any reason why this repo references a very old version of cloud ?
from go.mod, it's an explicit required module:
`google.golang.org/cloud v0.0.0-20151119220103-975617b05ea8`

my dependency chart shows:
`github.com/distribution/distribution/v3@v3.0.0-20220620080156-3e4f8a0ab147 google.golang.org/cloud@v0.0.0-20151119220103-975617b05ea8`

When building my tool that depends on this project, I get:
```
#22 61.60 /go/pkg/mod/google.golang.org/cloud@v0.0.0-20151119220103-975617b05ea8/storage/acl.go:65:16: invalid type assertion: v.(map[string]<inter>) (non-interface type *storage.ObjectAccessControl on left)
#22 61.60 /go/pkg/mod/google.golang.org/cloud@v0.0.0-20151119220103-975617b05ea8/storage/acl.go:144:16: invalid type assertion: v.(map[string]<inter>) (non-interface type *storage.ObjectAccessControl on left)
```

I'm running go 1.17.10

### PR DESCRIPTION
**Pull request by bshaaban**
This change updates google cloud dependency as the current version dates back to a 2015 version and does not compile with the latest Go version

closes #3668

Signed-off-by: Bahaiddine A <bshaaban@mirantis.com>

### CODE REVIEW

**Comment by thaJeztah** [comment]
Looks like there's an issue with the `go.sum` not matching the expected result;

```
#15 [validate 1/1] RUN --mount=target=/context   --mount=target=.,type=tmpfs <<EOT (set -e...)
#15 1.062 ERROR: Vendor result differs. Please vendor your package with "make vendor"
#15 1.233  M go.sum
```

**Comment by thaJeztah** [comment]
> and does not compile with the latest Go version

I'm curious though; do you have more details about this? I see CI on this repository is using Go 1.18; is it failing on 1.19beta1 or 1.18?

**Comment by bshaaban** [comment]
> > and does not compile with the latest Go version
> 
> I'm curious though; do you have more details about this? I see CI on this repository is using Go 1.18; is it failing on 1.19beta1 or 1.18?

actually i'm still on Go 1.17.10, I believe vendoring might not work the same based on the go version. Locally, `make vendor` works but `make validate-vendor` fails for me.

**Comment by thaJeztah** [comment]
Both should be running inside a Dockerfile, so afaics, the host environment should not make a difference; https://github.com/distribution/distribution/blob/26a586cf39831db7afab1ad97cc8556734919ffb/Makefile#L106-L114

Could it be you tried to run `make validate-vendor` before committing the changes?

**Comment by bshaaban** [comment]
> Both should be running inside a Dockerfile, so afaics, the host environment should not make a difference;
> 
> https://github.com/distribution/distribution/blob/26a586cf39831db7afab1ad97cc8556734919ffb/Makefile#L106-L114
> 
> Could it be you tried to run `make validate-vendor` before committing the changes?

this is probably right. I haven't run `make validate-vendor` prior to pushing my change. I've pushed an update to include this go.sum change. Probably additional changes are needed in the code base to update the google cloud storage calls to match their most recent updates/tags.

**Comment by bshaaban** [comment]
It seems the logic in `registry/storage/driver/gcs/gcs.go` must be refactored to take into account the new google cloud storage api, code compilation is skipped to to the go build tag in the file.

**Comment by Jamstah** [comment]
> It seems the logic in registry/storage/driver/gcs/gcs.go must be refactored to take into account the new google cloud storage api, code compilation is skipped to to the go build tag in the file.

Are you saying that there are additional changes required before users can compile for google cloud and use google cloud storage for their registry?

If so, I'd suggest making this a draft PR.

**Comment by bshaaban** [comment]
according to the comments in https://github.com/distribution/distribution/blob/main/registry/storage/driver/gcs/gcs.go#L317-L318, the logic was copied from an outdated google cloud storage api. I believe this logic needs to be updated as well.

**Comment by Jamstah** [comment]
Yes, I don't think the Google cloud storage driver is working. I expect it's not worth merging this PR until the work to fix it is done.

Are you planning to work on it?

**Comment by Jamstah** [comment]
I was thinking of doing that the other day:
> We should also list the available build-tags (and their meaning) under https://github.com/distribution/distribution/blob/main/BUILDING.md#optional-build-tags

See https://github.com/distribution/distribution/pull/3687

File: registry/storage/driver/gcs/gcs_test.go [hunk 1]
```diff
@@ -15,7 +15,6 @@ import (
 	"golang.org/x/oauth2"
 	"golang.org/x/oauth2/google"
 	"google.golang.org/api/googleapi"
-	"google.golang.org/cloud/storage"
```
Comment (kaovilai) [inline registry/storage/driver/gcs/gcs_test.go:18]: No replacement storage import here.

**Comment by kaovilai** [comment]
> we should have a job in CI that at least tries to compile the binaries with those build-flags set (so that we can see when things break)

@thaJeztah are you opening a PR for this?

**Comment by kaovilai** [comment]
I would be interested in helping with this PR. Let me know if you have something I can help with.

**Comment by Jamstah** [comment]
> I would be interested in helping with this PR. Let me know if you have something I can help with.

I think the important thing is getting the storage driver working. Until it's working it's not even worth turning it on in ci or changing dependencies.

**Comment by jmontleon** [comment]
I have a PR started at https://github.com/kaovilai/distribution/pull/1

It's a mess, but the only thing failing at the moment is:
```
----------------------------------------------------------------------
FAIL: /home/jason/Documents/src/github.com/konveyor/distribution/registry/storage/driver/testsuites/testsuites.go:395: DriverSuite.TestContinueStreamAppendSmall

/home/jason/Documents/src/github.com/konveyor/distribution/registry/storage/driver/testsuites/testsuites.go:396:
    suite.testContinueStreamAppend(c, int64(32))
/home/jason/Documents/src/github.com/konveyor/distribution/registry/storage/driver/testsuites/testsuites.go:419:
    c.Assert(curSize, check.Equals, int64(len(contentsChunk1)))
... obtained int64 = 0
... expected int64 = 32
```

Trying to figure out what's going wrong, but not really certain atm.

**Comment by kaovilai** [comment]
Keep an eye on #3702

**Review by milosgajdos — APPROVED** [review]
LGTM. @thaJeztah can you double check if I missed anything, please 🙇

**Comment by milosgajdos** [comment]
@bshaaban do you want to remove `WIP` from the title of this PR or do you still consider it to be a WIP?

**Comment by kaovilai** [comment]
@milosgajdos I don't think this PR is doing anything. Merging this will still result in broken GCS code.
So maybe close this and merge https://github.com/distribution/distribution/pull/3702 instead?

The CI is passing because BUILDTAGS hasn't been passed to `go build` as demonstrated by CI failure in https://github.com/distribution/distribution/pull/3703

**Comment by kaovilai** [comment]
This PR is still WIP [per comment](https://github.com/distribution/distribution/pull/3676#issuecomment-1180631040)

**Comment by milosgajdos** [comment]
Will mark this as draft.

**Comment by bshaaban** [comment]
I can rebase on top of #3702 to maintain dep updates since gcs was updated there already.

**Comment by bshaaban** [comment]
closing this pr, in favor of https://github.com/distribution/distribution/pull/3702

**Pull request closed without merging**